magma/types/decoupled.py

```diff
@@ -0,0 +1,167 @@
+        if not issubclass(T, Type):
+            raise TypeError(
+                f"ReadyValid[T] expected T to be a subclass of m.Type not {T}"
+            )
+        fields = {"valid": Out(Bit), "data": Out(T), "ready": In(Bit)}
+        return type(f"ReadyValid[{T}]", (ReadyValid, ), fields)
+
+
+class ReadyValid(Product, metaclass=ReadyValidKind):
+    """
+    An I/O Bundle containing 'valid' and 'ready' signals that handshake
+    the transfer of data stored in the 'data' subfield.
+
+    The base protocol implied by the directionality is that
+    the producer uses the interface as-is (outputs bits)
+    while the consumer uses the flipped interface (inputs bits).
+
+    For readability, we provide these helpers to avoid the confusion related to
+    In versus Out:
+    * m.Consume(m.ReadyValid[T])
+    * m.Product(m.ReadyValid[T])
+
+    T is the type of data to be wrapped in Ready/Valid
+
+    The actual semantics of ready/valid are enforced via the use of concrete
+    subclasses.
+    """
+    def fired(self):
+        """
+        Returns a bit that is high when ready and valid are both high
+        """
+        if self.is_consumer():
+            if not self.ready.driven():
+                raise ReadyValidException(
+                    "Cannot invoke fired on consumer when ready isn't driven"
+                )
+
+            return self.valid & self.ready.value()
+
+        assert self.is_producer()
+        if not self.valid.driven():
+            raise ReadyValidException(
+                "Cannot invoke fired on consumer when valid isn't driven"
+            )
+        return self.valid.value() & self.ready
+
+    def enq(self, value, when=True):
+        """
+        Produces the data `value` when `when` is high
+
+        Uses "last connect semantics", which allows multiple `enq` methods to
+        be invoked with different conditions with precedence given to the later
+        invocations.
+        """
+        if not self.is_producer():
+            raise TypeError("Cannot invoke enq on consumer")
+        if when is not True:
+            if not (self.valid.driven() and self.data.driven()):
+                raise ReadyValidException(
+                    "Cannot use enq when without default driver"
+                )
+            curr_valid = self.valid.value()
+            when |= curr_valid
+            self.valid.unwire(curr_valid)
+
+            curr_data = self.data.value()
+            value = m.mux([curr_data, value], when)
+            self.data.unwire(curr_data)
+        self.valid @= when
+        self.data @= value
+        return value
+
+    def no_enq(self, when=True):
+        """
+        Indicates no enqueue occurs.  Valid is set to false, and bits are
+        connected to 0 (TODO: Should be an unitialized wire (Don't Care))
+        """
+        if not self.is_producer():
+            raise TypeError("Cannot invoke no_enq on consumer")
+        if when is not True:
```

Comment:
This condition doesn't work because it raises this error when `isinstance(when, Type)`:

```
    def __bool__(self) -> bool:
3124>       raise NotImplementedError("Converting magma bit to bool not supported")
3125E       NotImplementedError: Converting magma bit to bool not supported
3126
3127magma/bit.py:191: NotImplementedError
```

Reply:
ah I see. Didn't realize `when` could be a signal. might be helpful to do a type annotation on `when` (maybe `typing.Union[bool, Type]`).

Of course `not x` and `x is not True` are not equivalent, but it might be confusing to read `is not True`. A type annotation could be a concise way to document that.